src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs

```diff
@@ -100,103 +117,477 @@ internal sealed partial class HttpConnectionPool
 
-            Http3Connection? http3Connection = Volatile.Read(ref _http3Connection);
+            _http3RequestQueue.PruneCompletedRequestsFromHeadOfQueue(this);
 
-            if (http3Connection != null)
+            // Determine if we can and should add a new connection to the pool.
+            int availableHttp3ConnectionCount = _availableHttp3Connections?.Count ?? 0;
+            bool willInject = availableHttp3ConnectionCount == 0 &&                         // No available connections
+                !_pendingHttp3Connection &&                                                 // Only allow one pending HTTP3 connection at a time
+                _http3RequestQueue.Count > 0 &&                                             // There are requests left on the queue
+                (_associatedHttp3ConnectionCount == 0 || EnableMultipleHttp3Connections) && // We allow multiple connections, or don't have a connection currently
+                _http3RequestQueue.RequestsWithoutAConnectionAttempt > 0;                   // There are requests we haven't issued a connection attempt for
+
+            if (NetEventSource.Log.IsEnabled())
+            {
+                Trace($"Available HTTP/3.0 connections: {availableHttp3ConnectionCount}, " +
+                    $"Pending HTTP/3.0 connection: {_pendingHttp3Connection}, " +
+                    $"Requests in the queue: {_http3RequestQueue.Count}, " +
+                    $"Requests without a connection attempt: {_http3RequestQueue.RequestsWithoutAConnectionAttempt}, " +
+                    $"Total associated HTTP/3.0 connections: {_associatedHttp3ConnectionCount}, " +
+                    $"Will inject connection: {willInject}.");
+            }
+
+            if (willInject)
+            {
+                _associatedHttp3ConnectionCount++;
+                _pendingHttp3Connection = true;
+
+                RequestQueue<Http3Connection>.QueueItem queueItem = _http3RequestQueue.PeekNextRequestForConnectionAttempt();
+                _ = InjectNewHttp3ConnectionAsync(queueItem); // ignore returned task
+            }
+        }
+
+        [SupportedOSPlatformGuard("linux")]
+        [SupportedOSPlatformGuard("macOS")]
+        [SupportedOSPlatformGuard("Windows")]
+        private async Task InjectNewHttp3ConnectionAsync(RequestQueue<Http3Connection>.QueueItem queueItem)
+        {
+            Debug.Assert(IsHttp3Supported());
+
+            if (NetEventSource.Log.IsEnabled()) Trace("Creating new HTTP/3 connection for pool.");
+
+            // Queue the remainder of the work so that this method completes quickly
+            // and escapes locks held by the caller.
+            await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
+
+            Http3Connection? connection = null;
+            Exception? connectionException = null;
+            Http3ConnectionWaiter waiter = (Http3ConnectionWaiter)queueItem.Waiter;
+            HttpAuthority authority = waiter.Authority;
```

Comment:
This may be an older authority.
How come we're passing it along instead of deciding on the latest value here?

Reply:
Done.